src/lib/panorama.ts

```diff
@@ -0,0 +1,245 @@
+import * as THREE from 'three';
+
+import { fetch } from './utils';
+
+const TAU = 2 * Math.PI;
+const geometryCache = {};
+
+let oldPano: THREE.Group | THREE.Mesh | null = null;
+let newPano: THREE.Group | THREE.Mesh | null = null;
+let panoFadeTime = performance.now();
+
+export function animatePano(scene: THREE.Scene) {
+	scene.children
+		.filter((child) => child.name === 'panorama')
+		.forEach((pano) => {
+			if (pano !== newPano && pano !== oldPano) scene.remove(pano);
+		});
+	if (newPano !== null) {
+		const now = performance.now();
+		const elapsed = now - panoFadeTime;
+		const t = Math.min(elapsed / 300, 1);
+
+		if (newPano instanceof THREE.Group) {
+			newPano.traverse((child) => {
+				if (child instanceof THREE.Mesh) {
+					child.material.opacity = t;
+				}
+			});
+		} else if (newPano instanceof THREE.Mesh) {
+			(newPano.material as THREE.Material[]).forEach((material) => {
+				material.opacity = t;
+			});
+		}
+
+		if (t >= 1) {
+			if (oldPano !== null) scene.remove(oldPano);
+			oldPano = newPano;
+			newPano = null;
+			if (oldPano instanceof THREE.Group) {
+				oldPano.traverse((child) => {
+					if (child instanceof THREE.Mesh) {
+						child.renderOrder = 0;
+					}
+				});
+			} else if (oldPano instanceof THREE.Mesh) {
+				oldPano.renderOrder = 0;
+			}
+		}
+	}
+}
+
+export function renderPanoFromId(
+	scene: THREE.Scene,
+	pmremGenerator: THREE.PMREMGenerator,
+	panoId: string,
+	gameHeading: number,
+	maxZoom: number,
+) {
+	if (panoId.startsWith('CAoS') && panoId.length > 22) {
+		const matches = atob(panoId.replaceAll('-', '+').replaceAll('_', '/').replaceAll('.', '=')).match(
+			/[A-Za-z0-9\-_]+$/,
+		);
+		panoId = matches !== null ? matches[0] : panoId;
+	}
+	const protobuf = panoId.startsWith('CIHM0og') || panoId.startsWith('AF1Qip') || panoId.length > 22;
+	fetch({
```

Comment:
Could just make the function async, right?

```suggestion
	await fetch({
```

Then any caller of it can also `await`/`.then()` it to know when the pano has finished rendering

Reply:
The original reason I'm not currently doing is because I wanted the .catch... But now that I think about it you can still just await and have a .catch at the same time so... yes. But if we're gonna do that I'd prefer assigning the value with `const res = await fetch(...` and surrounding it with a try catch. I like the goal of `await`/`.then()` to know when its finished rendering but I'd implement it differently.

Reply:
Did you have anything else in mind? How would you implement it differently?

Reply:
If my understanding is correct, *just* putting `await` on the fetch would only tell you when the metadata is done fetching not when the panorama is done rendering. Hence, this should be implemented differently.

Reply:
No, because you'd be putting an `await` on the already-`then()`ed Promise.

The simplest solution really is to put a `return` next to the fetch 😆

Reply:
Achieved a similar with the split of `renderPanoramaFromId` into `getPanoMetadataFromId` and `renderPanoFromMetadata` in 187588d08bb21976c2367051defce5cf7fa27073